File: cpp-linux/pdfa_extension/checkconformance/checkconformance.cpp
// checkconformance -- C++ port of examples\Vb6\pdfa_extension\checkconformance\checkconformance.bas
// Imports a PDF and converts it to PDF/A-3b via CheckConformance, using callbacks to
// replace missing fonts and ICC profiles, then writes the result.
#include <lumaspdf.h>
#include "repo_root.h"
#include <cstdio>
#include <string>

static const char* ICC_RGB  = LUMAS_REPO_ROOT "/sample_rgb.icc";
static const char* ICC_CMYK = LUMAS_REPO_ROOT "/sample_rgb.icc";
static const char* ICC_GRAY = LUMAS_REPO_ROOT "/sample_gray.icc";
static const char* INPUT    = LUMAS_REPO_ROOT "/sample_multipage.pdf";

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

// Data is the PDF handle passed as UserData to CheckConformance.
static SI32 PDF_CALL FontNotFoundProc(void* Data, void* PDFFont, const char* FontName, TFStyle Style, SI32 StdFontIndex, LBOOL IsSymbolFont){
    return pdfReplaceFontA((PPDF)Data, PDFFont, "Arial", (SI32)Style, 1);
}

static SI32 PDF_CALL ReplaceICCProfileProc(void* Data, TICCProfileType ProfileType, SI32 ColorSpace){
    switch(ProfileType){
        case ictRGB:  return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, ICC_RGB);
        case ictCMYK: return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, ICC_CMYK);
        default:      return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, ICC_GRAY);
    }
}

static bool ConvertFile(PPDF pdf, SI32 convType, const char* inFile, const char* outFile){
    pdfCreateNewPDFA(pdf, "");
    pdfSetDocInfoA(pdf, diProducer, "");

    UI32 convFlags;
    switch(convType){
        case ctNormalize:      convFlags = coAllowDeviceSpaces; break;
        case ctPDFA_1b_2005:   convFlags = coDefault | coFlattenLayers; break;
        case ctPDFA_2b:
        case ctPDFA_2u:        convFlags = coDefault | coDeletePresentation; break;
        default:               convFlags = (coDefault | coDeletePresentation) & ~coDeleteEmbeddedFiles; break;
    }
    convFlags |= coCheckImages;
    convFlags |= coRepairDamagedImages;

    if(convType != ctNormalize){
        pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage | ifPrepareForPDFA);
        pdfSetImportFlags2(pdf, if2UseProxy | if2DuplicateCheck);
    } else {
        pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage);
        pdfSetImportFlags2(pdf, if2UseProxy | if2DuplicateCheck | if2Normalize);
    }

    if(pdfOpenImportFileA(pdf, inFile, ptOpen, "") < 0){
        printf("Could not open the import file (it may be encrypted)!\n");
        pdfFreePDF(pdf);
        return false;
    }
    pdfImportPDFFile(pdf, 1, 1.0, 1.0);
    pdfCloseImportFile(pdf);

    SI32 retval = pdfCheckConformance(pdf, convType, convFlags, pdf, FontNotFoundProc, ReplaceICCProfileProc);
    switch(retval){
        case 1: pdfAddOutputIntentA(pdf, ICC_RGB);  break;
        case 2: pdfAddOutputIntentA(pdf, ICC_CMYK); break;
        case 3: pdfAddOutputIntentA(pdf, ICC_GRAY); break;
    }

    TPDFError e{};
    e.StructSize = sizeof(e);
    for(SI32 i = 0; i < pdfGetErrLogMessageCount(pdf); ++i){
        pdfGetErrLogMessage(pdf, i, &e);
        if(e.Msg) printf("%s\n", e.Msg);
    }

    if(pdfHaveOpenDoc(pdf) != 0){
        if(pdfOpenOutputFileA(pdf, outFile) == 0) return false;
        return pdfCloseFile(pdf) != 0;
    }
    return false;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, PDFError);

    std::string cmapDir = exeDir(argv[0]) + "/CMap";
    pdfSetCMapDirA(pdf, cmapDir.c_str(), lcmDelayed | lcmRecursive);
    std::string filePath = exeDir(argv[0]) + "/out.pdf";

    if(ConvertFile(pdf, ctPDFA_3b, INPUT, filePath.c_str()))
        printf("PDF file \"%s\" successfully created!\n", filePath.c_str());

    pdfDeletePDF(pdf);
    return 0;
}

File: vbnet/pdfa_extension/checkconformance/checkconformance.vb
' checkconformance -- VB.NET port of examples\Vb6\pdfa_extension\checkconformance
Imports System
Imports System.Runtime.InteropServices
Imports LumasPdfSdk

Module modCheckConformance
    Private errDel As TErrorProc
    Private fontDel As TOnFontNotFoundProc
    Private iccDel As TOnReplaceICCProfile

    Private Function U(ByVal v As Long) As UInteger
        Return CUInt(v And &HFFFFFFFFL)
    End Function

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Public Function PDFError(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return 0
    End Function

    Public Function FontNotFoundProc(ByVal Data As IntPtr, ByVal PDFFont As IntPtr, ByVal FontName As String, ByVal Style As Integer, ByVal StdFontIndex As Integer, ByVal IsSymbolFont As Boolean) As Integer
        Return LumasPdf.pdfReplaceFontW(Data, PDFFont, "Arial", Style, True)
    End Function

    Public Function ReplaceICCProfileProc(ByVal Data As IntPtr, ByVal ProfileType As TICCProfileType, ByVal ColorSpace As Integer) As Integer
        Select Case ProfileType
            Case TICCProfileType.ictRGB
                Return LumasPdf.pdfReplaceICCProfileW(Data, U(ColorSpace), "../../../test_files/sRGB.icc")
            Case TICCProfileType.ictCMYK
                Return LumasPdf.pdfReplaceICCProfileW(Data, U(ColorSpace), "../../../test_files/ISOcoated_v2_bas.ICC")
            Case Else
                Return LumasPdf.pdfReplaceICCProfileW(Data, U(ColorSpace), "../../../test_files/gray.icc")
        End Select
    End Function

    Private Function ConvertFile(ByVal pdf As IntPtr, ByVal ConvType As Integer, ByVal InFile As String, ByVal OutFile As String) As Boolean
        ConvertFile = False
        LumasPdf.pdfCreateNewPDFW(pdf, "")
        LumasPdf.pdfSetDocInfoW(pdf, TDocumentInfo.diProducer, "")

        Dim convFlags As Integer
        Select Case ConvType
            Case CInt(TConformanceType.ctNormalize)
                convFlags = LumasPdfConsts.coAllowDeviceSpaces
            Case CInt(TConformanceType.ctPDFA_1b_2005)
                convFlags = LumasPdfConsts.coDefault Or LumasPdfConsts.coFlattenLayers
            Case CInt(TConformanceType.ctPDFA_2b), CInt(TConformanceType.ctPDFA_2u)
                convFlags = LumasPdfConsts.coDefault Or LumasPdfConsts.coDeletePresentation
            Case Else
                convFlags = (LumasPdfConsts.coDefault Or LumasPdfConsts.coDeletePresentation) And (Not LumasPdfConsts.coDeleteEmbeddedFiles)
        End Select
        convFlags = convFlags Or LumasPdfConsts.coCheckImages
        convFlags = convFlags Or LumasPdfConsts.coRepairDamagedImages

        If ConvType <> CInt(TConformanceType.ctNormalize) Then
            LumasPdf.pdfSetImportFlags(pdf, U(LumasPdfConsts.ifImportAll Or LumasPdfConsts.ifImportAsPage Or LumasPdfConsts.ifPrepareForPDFA))
            LumasPdf.pdfSetImportFlags2(pdf, U(LumasPdfConsts.if2UseProxy Or LumasPdfConsts.if2DuplicateCheck))
        Else
            LumasPdf.pdfSetImportFlags(pdf, U(LumasPdfConsts.ifImportAll Or LumasPdfConsts.ifImportAsPage))
            LumasPdf.pdfSetImportFlags2(pdf, U(LumasPdfConsts.if2UseProxy Or LumasPdfConsts.if2DuplicateCheck Or LumasPdfConsts.if2Normalize))
        End If

        If LumasPdf.pdfOpenImportFileW(pdf, InFile, LumasPdfConsts.ptOpen, "") < 0 Then
            Console.WriteLine("Could not open the import file (it may be encrypted)!")
            LumasPdf.pdfFreePDF(pdf)
            Return False
        End If
        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0)
        LumasPdf.pdfCloseImportFile(pdf)

        Dim retval As Integer = LumasPdf.pdfCheckConformance(pdf, ConvType, U(convFlags), pdf, fontDel, iccDel)
        Select Case retval
            Case 1 : LumasPdf.pdfAddOutputIntentW(pdf, "../../../test_files/sRGB.icc")
            Case 2 : LumasPdf.pdfAddOutputIntentW(pdf, "../../../test_files/ISOcoated_v2_bas.ICC")
            Case 3 : LumasPdf.pdfAddOutputIntentW(pdf, "../../../test_files/gray.icc")
        End Select

        Dim e As TPDFError
        e.StructSize = CUInt(Marshal.SizeOf(GetType(TPDFError)))
        For i As Integer = 0 To LumasPdf.pdfGetErrLogMessageCount(pdf) - 1
            LumasPdf.pdfGetErrLogMessage(pdf, CUInt(i), e)
            Console.WriteLine(Marshal.PtrToStringAnsi(e.Msg))
        Next

        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            If LumasPdf.pdfOpenOutputFileW(pdf, OutFile) = 0 Then
                Return False
            End If
            ConvertFile = (LumasPdf.pdfCloseFile(pdf) <> 0)
        End If
    End Function

    Sub Main()
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        errDel = New TErrorProc(AddressOf PDFError)
        fontDel = New TOnFontNotFoundProc(AddressOf FontNotFoundProc)
        iccDel = New TOnReplaceICCProfile(AddressOf ReplaceICCProfileProc)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)

        LumasPdf.pdfSetCMapDirW(pdf, AppPath() & "\..\..\..\Resource\CMap", U(LumasPdfConsts.lcmDelayed Or LumasPdfConsts.lcmRecursive))
        Dim filePath As String = AppPath() & "\out.pdf"

        If ConvertFile(pdf, CInt(TConformanceType.ctPDFA_3b), AppPath() & "\sample_multipage.pdf", filePath) Then
            Console.WriteLine("PDF file """ & filePath & """ successfully created!")
        End If

        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module
